lib/service/connect.go

```diff
@@ -67,24 +67,24 @@ func (process *TeleportProcess) reconnectToAuthService(role types.SystemRole) (*
 	var assertionID string
 
 	for {
-		connector, err := process.connectToAuthService(role, systemRoleAssertionID(assertionID))
-		if err == nil {
+		connector, connectErr := process.connectToAuthService(role, systemRoleAssertionID(assertionID))
+		if connectErr == nil {
 			// if connected and client is present, make sure the connector's
 			// client works, by using call that should succeed at all times
 			if connector.Client != nil {
-				pingResponse, err := connector.Client.Ping(process.ExitContext())
-				compareErr := process.authServerTooOld(&pingResponse)
-				if compareErr != nil {
-					return nil, trace.Wrap(compareErr)
-				}
+				pingResponse, pingErr := connector.Client.Ping(process.ExitContext())
+
+				if pingErr == nil {
```

Comment:
The error checks are backward here because we want to return early if everything passed, and continue down the loop body if there are errors.

I removed one level of `if` (client should never be nil if `connectErr` is nil) and added a bunch of comments to hopefully make things more clear.

> Should we allow things to proceed if the ping failed and we can't validate the version?

- If the ping failed, we definitely proceed with the loop and try to get a new, working client.
  - previously we would check the empty version from the failed ping response in this case, but that doesn't make sense
 - If the ping succeeded:
   - if the version check fails, it will probably continue to fail on every loop, so we return an error early
   - if the version check passed, good, we have a working client and return it